### PR TITLE
Moved all API calls to new internal functions

### PR DESCRIPTION
**Pull request by ecspresso**
Added `New-ITGlue`
Added `Get-ITGlue`
Added `Set-ITGlue`
Added `Remove-ITGlue`

All GET requests have their filter parameters moved to the `$filter_list` variable. The internal function `Get-ITGlue` handles creating the new URI.

All other requests with filters have their filter parameters moved from the `$body` variable also to the `$filter_list` variable. The internal function will still add all filters to the body (but this can easily be changed, **take special note of show_password in Passwords.ps1 if we do, the url is still handled by the external function as a result of this**) because that is how it was documented.

`$body` variable has been removed from all exposed functions and move to the internal `*-ITGlue` functions.

### CODE REVIEW

**Comment by ecspresso** [comment]
I don't have access to IT Glue anymore, so run some test before any merge is done.

**Comment by ecspresso** [comment]
This PR addresses #151.

**Comment by ecspresso** [comment]
As stated in the first comment, `set` and `remove` still use filters in their body of the request because that is how it is documented. All we need to if we decide to change this to copy how it is done in the `get` function, the filter parameters are kept seperate but still added to the body.

I should also add that I changed to proper hashtable handling (i.e. `filter_list['filter[id]'] = $filter_id` instead of   `filter_list += @{'filter[id]'=$filter_id}`).

File: ITGlueAPI/ITGlueAPI.psd1 [hunk 1]
```diff
@@ -18,7 +18,7 @@
     # -- MINOR version when you add functionality in a backwards-compatible manner, and
     # -- PATCH version when you make backwards-compatible bug fixes.
 
-    ModuleVersion = '2.2.0'
+    ModuleVersion = '2.1.0'
```
Comment (davidhaymond) [inline ITGlueAPI/ITGlueAPI.psd1:21]: The version has been bumped to 2.2.0 in upstream master, so don't forget to update your local branch to incorporate these changes.

File: ITGlueAPI/Internal/APICalls.ps1 [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+    try {
+        $ITGlue_Headers.Add('x-api-key', (New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList 'N/A', $ITGlue_API_Key).GetNetworkCredential().Password)
+
+        $parameters = @{
+            "Method" = 'GET'
+            "Uri" = $ITGlue_Base_URI + $resource_uri
+            "Headers" = $ITGlue_Headers
+        }
+
```
Comment (davidhaymond) [inline ITGlueAPI/Internal/APICalls.ps1:61]: There's still a lot of code duplication across the four API call functions. I think they should be merged into a single `Invoke-APIRequest` (or similar) function. Here's my suggestion:

```powershell
function ConvertTo-QueryString([Hashtable]$url_params) {
    if ($null -eq $url_params) {
        return ""
    }

    $params = @()
    foreach ($key in $url_params.Keys) {
        $params += "$key=$($url_params[$key])"
    }

    $query_string = '?' + ($params -join '&')
    return $query_string
}

function Invoke-APIRequest {
    param (
        [ValidateSet('GET', 'POST', 'PATCH', 'DELETE')]
        [string]$method = 'GET',

        [Parameter(Mandatory = $true)]
        [string]$resource_uri,

        [Hashtable]$url_params = $null,

        [Hashtable]$data = $null
    )

    $query_string = ConvertTo-QueryString -url_params $url_params

    if ($null -eq $data) {
        $body = $null
    } else {
        $body = @{'data'=$data} | ConvertTo-Json -Depth $ITGlue_JSON_Conversion_Depth
    }

    try {
        $ITGlue_Headers.Add('x-api-key', (New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList 'N/A', $ITGlue_API_Key).GetNetworkCredential().Password)

        $parameters = @{
            "Method" = $method
            "Uri" = $ITGlue_Base_URI + $resource_uri + $query_string
            "Headers" = $ITGlue_Headers
            "Body" = $body
        }

        $api_response = Invoke-RestMethod @parameters -ErrorAction Stop
    } catch {
        Write-Error $_
    } finally {
        [void] ($ITGlue_Headers.Remove('x-api-key')) # Quietly clean up scope so the API key doesn't persist
    }

    return $api_response
}
```

Of course, the calls to this function in the resource files would all need to be updated accordingly.

File: ITGlueAPI/Resources/Passwords.ps1 [hunk 3]
```diff
@@ -176,25 +155,7 @@ function Set-ITGluePasswords {
         $resource_uri = $resource_uri + ('?show_password=true') # using $True in PS results in 'True' (uppercase T), so explicitly writing out 'true' is needed
```
Comment (davidhaymond) [inline ITGlueAPI/Resources/Passwords.ps1:155]: `show_password` shouldn't go in $resource_uri; it should be parsed as a query string parameter and added by the internal function.

**Review by davidhaymond — CHANGES_REQUESTED** [review]
Wow, you've put a lot of work into this, @ecspresso! I think there's more cleanup to do (see my comments below), but this is a fantastic start!

File: ITGlueAPI/Resources/Passwords.ps1 (hunk 3 above)
Comment (davidhaymond) [reply ITGlueAPI/Resources/Passwords.ps1:155]: I did some testing and confirmed that `show_password` only works in the query string, even for POST or PATCH requests.

**Comment by davidhaymond** [comment]
Since I haven't heard anything from @ecspresso, I went ahead and incorporated my suggestions into #160. This PR should be closed. @CalebAlbers

**Comment by adrianwells** [comment]
As @davidhaymond noted, closing this PR as its efforts are in #160. Thank you @ecspresso for the work here.

**Pull request closed without merging**